pandas/_testing.py

```diff
@@ -2641,18 +2643,37 @@ class for all warnings. To check that no warning is returned,
                         actual_warning.lineno,
                     )
                 )
-        if expected_warning:
-            msg = (
+
+        if expected_warning and not saw_warning:
+            raise AssertionError(
                 f"Did not see expected warning of class "
                 f"{repr(expected_warning.__name__)}"
             )
-            assert saw_warning, msg
+
+        if match and not matched_message:
+            raise AssertionError(
+                f"Did not see warning {repr(expected_warning.__name__)} "
+                f"matching {match}"
+            )
+
         if raise_on_extra_warnings and extra_warnings:
             raise AssertionError(
                 f"Caused unexpected warning(s): {repr(extra_warnings)}"
             )
 
 
+def _assert_raised_with_correct_stacklevel(actual_warning):
```

Comment:
bonus points for typing

Reply:
Annotated. IMHO, the type ``warnings.WarningMessage``, although correct, has a confusing name for warning instance.